### PR TITLE
Issue 7047 optional jemalloc

### LINKED ISSUE #7047: Optionally load jemalloc
### Description
As an Developer/Devops/Tester, I want to enable/disable the use of jemalloc when starting Besu so that I can easily test performance with/without jemalloc and be able to start without jemalloc if there is a memory issue on startup.

### Acceptance Criteria
* Besu will load jemalloc optionally based on a env var

### Versions
* All Linux based versions

### Implementation

I was able to implement and test this feature by reusing the `BESU_USING_JEMALLOC` env var, with changes in the initial _besu start script_ and a few lines in _ConfigurationOverviewBuilder_. 

If set, `BESU_USING_JEMALLOC=true` will load jemalloc, if set to anything else it won't. If `BESU_USING_JEMALLOC`  is not set at all it'll load it too (that is the existing behavior). 

If enough people think this is useful I am ready to create a PR.

### PR DESCRIPTION
**Pull request by amsmota**
## Fixed Issue(s)

Fixes #7047 

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

File: besu/src/main/scripts/unixStartScript.txt
Comment (usmansaleem) [inline besu/src/main/scripts/unixStartScript.txt:185]: @amsmota I believe the original intent of this section of code is to detect availability of jemalloc library on the system and then set BESU_USING_JEMALLOC variable automatically. In other words, after you commented out the line 185 and 193, all of this section becomes redundant. 

What was the original problem that you have observed and trying to solve?

**Review by usmansaleem — DISMISSED** [review]
See the comment please.

File: CHANGELOG.md
Comment (usmansaleem) [inline CHANGELOG.md:165]: Also, this section needs to be moved to the top of changelog.

File: besu/src/main/scripts/unixStartScript.txt
Comment (usmansaleem) [reply besu/src/main/scripts/unixStartScript.txt:185]: In fact, regardless of the environment variable is present or not (or calculated by the `unixStartScript` or not), The Besu code will always attempt to detect jemalloc for logging purposes. 

https://github.com/hyperledger/besu/blob/8ca7129b0be3ba26b60ceb8920c01f717d5f5aca/besu/src/main/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilder.java#L431

As I see it, the code in `detectJemalloc` should be simplified. The main purpose of `detectJeMalloc` is to simply **log** that the jemalloc library is being used by Besu Java process. The main issue that I see is our `unixStartScript` is the setting of `LD_PRELOAD`, it is possible that it is not detecting the correct path. 

This is why I am interested to know exactly what issue you are facing?

Comment (usmansaleem) [reply besu/src/main/scripts/unixStartScript.txt:185]: Note that the purpose of environment variable `BESU_USING_JEMALLOC` is not to enable/disable using jemalloc. It is mainly used as an environment variable controlling the informational logging when Besu starts. The main variable that controls loading of jemalloc and make it available to Besu Java process is `LD_PRELOAD`.

Comment (usmansaleem) [reply besu/src/main/scripts/unixStartScript.txt:185]: @fab-10 is my understanding about BESU_USING_JEMALLOC correct? ^

**Comment by fab-10** [comment]
@amsmota @usmansaleem I just had a different idea about how to implement, that is more in line with the requirements. Basically we have the `besu-untuned` script that seems to perfectly fit the purpose of having a _vanilla_ Besu command, without any special options enforced, that can be used to perform any kind of customization, so I would rather suggest to keep `besu` command as is, and remove all the `jemalloc` related code from `besu-untuned`.

File: besu/src/main/scripts/unixStartScript.txt
Comment (fab-10) [reply besu/src/main/scripts/unixStartScript.txt:185]: correct

Comment (amsmota) [reply besu/src/main/scripts/unixStartScript.txt:185]: Guys, thank you for reviwing and your comments. 

The intention of this PR is to be able for the developers or testers or DevOps people to be able to load or not load jemalloc based on the BESU_USING_JEMALLOC env. The motivation for this are:

- Developpers and Testers to be able to quickly compare the improvement in memory resources to be able to have a informed decision on using Jemalloc or another memory allocator or none at all.
- DevOps guys to be able to disable Jemalloc if the deployement fails. A real world scenario would be for the deployment to be sucessfull in DEV and UAT and fail in PROD (maybe because sligthly diferences in OS version or memory configuration or security settings, etc...). With this optional flag they can simply set BESU_USING_JEMALLOC to false and restart the deployment. Without it they would have no alternative to rollback the deployement and probably send that back to the dev team to remove Jemalloc altogether.

In relation to the BESU_USING_JEMALLOC flag, the intention is that it will be set at the operation system level, not in the script. That is why I commented out the lines in the unixStartScript. I left them there commented out there so reviwers could see what I did and eventually test for thenselves with or without it. 

The ConfigurationOverviewBuilder handles all the cases:

1) BESU_USING_JEMALLOC == true, it will load Jemalloc if it's not loaded yet:
ConfigurationOverviewBuilder 
```
                if (Boolean.parseBoolean(jemallocEnabled)) {
                  final String version = PlatformDetector.getJemalloc();
                  lines.add("jemalloc: " + version);
                } else {
                  logger.info(
                      "BESU_USING_JEMALLOC is present but is not set to true, jemalloc library not loaded");
                }
```
PlatformDetector
```
  public static String getJemalloc() {
    if (_jemalloc == null) {
      detectJemalloc();
    }

    return _jemalloc;
  }
```
```
 private static void detectJemalloc() {
    interface JemallocLib extends Library {...}

   *****************************************************************************
    final JemallocLib jemallocLib = Native.load("jemalloc", JemallocLib.class);
   *****************************************************************************

    PointerByReference pVersion = new PointerByReference();
    IntByReference pSize = new IntByReference(Native.POINTER_SIZE);
    jemallocLib.mallctl("version", pVersion, pSize, null, 0);

    _jemalloc = pVersion.getValue().getString(0);
}
```

2) BESU_USING_JEMALLOC == false, it will NOT load, it'll do nothing except
```
                  logger.info(
                      "BESU_USING_JEMALLOC is present but is not set to true, jemalloc library not loaded")
```
3) BESU_USING_JEMALLOC  does not exist at all, it will behave like the current implementatio (tries to load anyway)
```
              // in case the user is using jemalloc without BESU_USING_JEMALLOC env var
              try {
                final String version = PlatformDetector.getJemalloc();
                lines.add("jemalloc: " + version);
              } catch (final Throwable throwable) {
                logger.info(
                    "jemalloc library not found, memory usage may be reduced by installing it");
              }
```

I considered to change this last option and not load Jemalloc when the env var does not exist. It's probably safer but I decided to keep it compatible with the current impelmentation.

I tested this quite intensivelly with all 3 scenarios, I actually installed WSL on my Widnows machine to do it.

Please let me know what you think.

Cheers.

Comment (fab-10) [reply besu/src/main/scripts/unixStartScript.txt:185]: Hi @amsmota, the requirements are clear, and I think the best way to implement them is to work on `besu-untuned` to completely remove the support for `jemalloc` from it, so you do not have any pre-configured tuning, and you can specify only the options you need, see my previous comment https://github.com/hyperledger/besu/pull/7424#issuecomment-2284458901

**Comment by amsmota** [comment]
> @amsmota @usmansaleem I just had a different idea about how to implement, that is more in line with the requirements. Basically we have the `besu-untuned` script that seems to perfectly fit the purpose of having a _vanilla_ Besu command, without any special options enforced, that can be used to perform any kind of customization, so I would rather suggest to keep `besu` command as is, and remove all the `jemalloc` related code from `besu-untuned`.

Hi, TBH I don't know what that besu-untuned is, do you have any link to docs or something? Anyway, besu already supports Jemalloc, right? They did intensive tests with several mallocs and they found out Jemalloc is the one that gives better performance, an it is part of the product now. Are they supposed to remove it from besu main? If not this PR fits the bill, it contains minimal changes to it and it works fine. So I would like to go ahead with this PR.

https://wiki.hyperledger.org/display/BESU/2024+-+Besu+Performance+Improvements+since+the+Merge
https://wiki.hyperledger.org/display/BESU/Reduce+Memory+usage+by+choosing+a+different+low+level+allocator

On a personal note, I won't have much time to contribute to Besu since I moved from Blockchain to GenAI and I have all my time assigned to it, that's why I want to wrap up my 2 PRs as fast as I can...

Cheers.

File: besu/src/main/scripts/unixStartScript.txt
Comment (amsmota) [reply besu/src/main/scripts/unixStartScript.txt:185]: > Note that the purpose of environment variable `BESU_USING_JEMALLOC` is not to enable/disable using jemalloc. It is mainly used as an environment variable controlling the informational logging when Besu starts. The main variable that controls loading of jemalloc and make it available to Besu Java process is `LD_PRELOAD`.

It is not so, in PlatformDetector::detectJemalloc it does load Jemalloc as I posted in my other msg...

**Comment by fab-10** [comment]
@amsmota `besu-untuned` is a start script, you can find it in the `bin` dir, that is like the `besu` standard script, but hasn't any preset configuration for Java memory, so you can use it to test your own setup, for example change GC, or other flags, that is exactly what you want to test a custom setup, to see if it is better than the proposed one. 
So removing the preset of the allocator from `besu-untuned` goes in the right directing of giving the freedom to the user to selected whichever allocator he want, and should have been already the case.
Do not worry, I can pick this task if you haven't time now, and thanks for raising that requirement.

**Comment by amsmota** [comment]
> @amsmota `besu-untuned` is a start script, you can find it in the `bin` dir

Oh, I saw it now... I thougth it was a fork of the main project or so...  I never care to look at that folder after installDist... So with that "untuned" version it make all sense to remove any reference to Jemalloc as you say, and let it use the OS default Malloc and let the users/developers to do what they think it's better. However, the fact is that the main Besu already support that, for what I see since 22.7.0-RC2, so I think it's still good to provied this change to the "general" population that uses Besu out-of-the-box. In our case this concerns more the Prod deployment, because we need to quickly restart a deployment without Jemalloc if there are failures, and the current code will always load Jemalloc if it is available...

In the meantime I updated my branch to the latest main, I think I'll have to test everithing again...

Cheers.

**Comment by fab-10** [comment]
To use `jemalloc` with the `besu-untuned` script (when it will be changed), you just need to run
```
export LD_PRELOAD=libjemalloc.so; bin/besu-untuned
```

File: besu/src/main/scripts/unixStartScript.txt
Comment (fab-10) [reply besu/src/main/scripts/unixStartScript.txt:185]: `PlatformDetector::detectJemalloc` is only needed to fetch the version of `jemalloc` on the overview, it makes no difference if that code is run or not.
The way `jemalloc` is used by the process is via `LD_PRELOAD`

**Comment by amsmota** [comment]
Ok, guys, I updated and build my branch, I uninstalled jemalloc, tested with BESU_USING_JEMALLOC true and false and without it, all messages appeared correctly, I installed jemalloc again and tested with BESU_USING_JEMALLOC true and false and without it and all messages appeared correctly too. 

I also tested removing all LD_PRELOAD from the scripts and you guys are right, it is needed to load jemalloc, my bad... It's not that `Native.load` won't load, it loads it but just for Java internal calls, but it does not influence the behavior of memory allocation.

I also tested with LD_PRELOAD but removing the `Native.load` part, it works fine but I have no way to confirm that jemalloc is loaded. However, I have no reason to believe it dosen't, so I assume it is loaded.

So what I am going to do is to remove the BESU_USING_JEMALLOC from the script, since it does not exist it will load jemalloc if exists any way, and let users set BESU_USING_JEMALLOC if and as they need it...

What do youse guys think?

**Comment by amsmota** [comment]
> To use `jemalloc` with the `besu-untuned` script (when it will be changed), you just need to run
> 
> ```
> export LD_PRELOAD=libjemalloc.so; bin/besu-untuned
> ```

Yes, the only issue is that there will be nothing in the logs saying it is using it. But then again, whoever adds the LD_PRELOAD will know. It will be a nice-to-have to log some info, but will not change any behaviour...

File: besu/src/main/scripts/unixStartScript.txt
Comment (amsmota) [reply besu/src/main/scripts/unixStartScript.txt:185]: Yep, that's right, my bad... :(

**Review by fab-10 — APPROVED** [review]

File: CHANGELOG.md
Comment (usmansaleem) [inline CHANGELOG.md:16]: This changelog entry needs to change specially if we simplify detectMalloc method. This can be rephrased similar to: "Simplified logging of jemalloc library at Besu startup".

**Review by usmansaleem — CHANGES_REQUESTED** [review]
The `detectJeMalloc` in `ConfigurationOverviewBuilder` can be simplified. It can also be renamed to `detectAndLogJeMallocVersion` to make the intent more clear.

**Comment by amsmota** [comment]
Guys, I really got this wrong, I assumed things that I shouldn't have assumed, like assuming that LD_PRELOAD and Native.load would both load the lib, which is true but they are used for diferent ends. Thank you guys for correcting me. However, my goal of be able to load or not load Jemmaloc when it does exist it's not achivable with any of the solutions I gave, so now what I think it should be done to achive that is change the unix script to be like 

```
if [ -z "$BESU_USING_JEMALLOC" ] || { [ -n "$BESU_USING_JEMALLOC" ] && [ "$BESU_USING_JEMALLOC" != "FALSE" ] && [ "$BESU_USING_JEMALLOC" != "false" ]; }; then
    if [ "\$darwin" = "false" -a "\$msys" = "false" ]; then
      # check if jemalloc is available
      TEST_JEMALLOC=\$(LD_PRELOAD=libjemalloc.so sh -c true 2>&1)

      # if jemalloc is available the output is empty, otherwise the output has an error line
      if [ -z "\$TEST_JEMALLOC" ]; then
        export LD_PRELOAD=libjemalloc.so
      else
        # jemalloc not available, as fallback limit malloc to 2 arenas
        export MALLOC_ARENA_MAX=2
      fi
    fi
fi
```

or probably with that IF just before the export LD_PRELOAD.

I did this change but I'm not able to test it, for some starnge reason both in Win and WSL I'm getting this error that I never saw, never happened before, and googling gives me no idea. I tried the obvious, to delete the entire build dir and add 777 access to subdirs, but it keeps on giving that error.

> * What went wrong:
> Execution failed for task ':evmToolStartScripts'.
> > Could not write to file 'C:\Users\anton\Development\OSPO\besu\build\scripts\evmtool'

Anyway, I'll commit in case any of youse are able to test... I'll be able to do some work/testing tomorrow and Monday, probably I'm going to delete my local and checkout again, and test properly. And I'll check your latest comment too.

Thank you guys for all the help.

Cheers.

**Comment by fab-10** [comment]
@amsmota use this code, since this is a template for the shell script, you need to escape `$` and it is possible to simplify the test expression

```
if [ "\$darwin" = "false" -a "\$msys" = "false" ]; then
    if [ "\$BESU_USING_JEMALLOC" != "FALSE" -a "\$BESU_USING_JEMALLOC" != "false" ]; then
      # check if jemalloc is available
      TEST_JEMALLOC=\$(LD_PRELOAD=libjemalloc.so sh -c true 2>&1)

      # if jemalloc is available the output is empty, otherwise the output has an error line
      if [ -z "\$TEST_JEMALLOC" ]; then
        export LD_PRELOAD=libjemalloc.so
        export BESU_USING_JEMALLOC=true
      else
        # jemalloc not available, as fallback limit malloc to 2 arenas
        export MALLOC_ARENA_MAX=2
      fi
    fi
fi
```

**Comment by usmansaleem** [comment]
@amsmota See https://github.com/hyperledger/besu/pull/7424#issuecomment-2320674559 please.

**Comment by amsmota** [comment]
Hi @usmansaleem, @fab-10, I finished today my other PR so I'll be looking at this one. If you have further comments besides the ones above please let me know.

Cheers.

**Comment by amsmota** [comment]
> @amsmota See [#7424 (comment)](https://github.com/hyperledger/besu/pull/7424#issuecomment-2320674559) please.

Hi @fab-10, I saw it now, I'm sorry for the delay but I've been busy at work and also with my other PR. It looks good to me, yes. I'll try to test in my local and let youse know.

Cheers.

**Comment by amsmota** [comment]
Guys, again I'm sorry for the delays, I did changed my code with some changes requested but I have issues with my Windows 11 / WSL, specifically with Gradle in WSL. I may opt to push the code as it is and ask youse reviewers to test yourselves if I can’t fix that WSL.

Cheers.

**Comment by amsmota** [comment]
@usmansaleem, @fab-10, I deleted everithing from my WSL, both the code and the ~/.gradle. I then run a ./gradlew build but it fails:

> > Task :evmToolStartScripts FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':evmToolStartScripts'.
> > Could not write to file '/home/amsmota/Development/besu/build/scripts/evmtool'.

I am running with my user user, not with sudo, maybe I should try to do it as su?

Anyway, did youse saw this issue before and/or know how to fix it?

Cheers.

**Comment by amsmota** [comment]
> Could not write to file '/home/amsmota/Development/besu/build/scripts/evmtool'.

Well, I fixed this one 👍 , which is good because it was I that caused...... 👎. And BTW the fix was actually the @fab-10 code from https://github.com/hyperledger/besu/pull/7424#issuecomment-2320674559. I could swear I had done that at some point but it seems I didn't... 

But anyway, I have another error, one that I had before when working in https://github.com/hyperledger/besu/pull/6965 and I couldn't solve, and it is is a linux error too...

> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)

This doesen't happen building on windows, only in WSL.

There too some lines that I don't know of are related, I'll try to investigate that.

> Java HotSpot(TM) 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended

Cheers.

**Comment by amsmota** [comment]
@fab-10 , @usmansaleem  I finnally was able to build my branch and tested it and IT'S ALL GOOD !!!!

This are the tests I did and the results output in the log:

> NO Jemalloc Installed / NO BESU_USING_JEMALLOC
> jemalloc library not found, memory usage may be reduced by installing it
> 
> NO Jemalloc Installed / BESU_USING_JEMALLOC = false
> BESU_USING_JEMALLOC is present but is not set to true, jemalloc library not loaded
> 
> NO Jemalloc Installed / BESU_USING_JEMALLOC = true
> BESU_USING_JEMALLOC is present but we failed to load jemalloc library to get the version
> 
> 
> WITH Jemalloc Installed / NO BESU_USING_JEMALLOC
> jemalloc: 5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756
> 
> WITH Jemalloc Installed / BESU_USING_JEMALLOC = false
> BESU_USING_JEMALLOC is present but is not set to true, jemalloc library not loaded
> 
> WITH Jemalloc Installed / BESU_USING_JEMALLOC = true
> jemalloc: 5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756

I still have some work to do, namelly
- fix some jemalloc tests that are now failing
- change those outputs above to print BESU_USING_JEMALLOC  in lowers (I don't like the way they look in capitals in the log)
"besu_using_jemalloc is present but we failed to load jemalloc library to get the version"
- make a few changes in the script

A little concern I had is that I installed jemalloc on Ubuntu using the _libjemalloc2_ lib, that installs it as _libjemalloc.so.2_ . I had to rename it _libjemalloc.so_ to make it compatible with the script, a note somewhere should warn about this...

Anyhow, by tomorrow or Saturday I should have this pushed to the PR.

Cheers.

File: besu/src/main/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilder.java
Comment (amsmota) [reply besu/src/main/java/org/hyperledger/besu/cli/ConfigurationOverviewBuilder.java:434]: I'm sorry but I don't understand this one, BESU_USING_JEMALLOC is still in the script

if [ "\$BESU_USING_JEMALLOC" != "FALSE" -a "\$BESU_USING_JEMALLOC" != "false" ]; then

and may or not exist as a env var, actually that situation is expected in the current code

            () -> {
              // in case the user is using jemalloc without BESU_USING_JEMALLOC env var
              try {
                final String version = PlatformDetector.getJemalloc();
                lines.add("jemalloc: " + version);
              } catch (final Throwable throwable) {
                logger.info(
                    "jemalloc library not found, memory usage may be reduced by installing it");
              }
            });

So now the detecJemalloc outputs one of 3 warnings:

"besu_using_jemalloc is present but is not set to true, jemalloc library not loaded"
"besu_using_jemalloc is present but we failed to load jemalloc library to get the version"
"jemalloc library not found, memory usage may be reduced by installing it"

Actually, with this version now if no action is taken by the clients (meaning they do not create a BESU_USING_JEMALLOC enc var, like running it out of the box) the code that actually runs is that one above that exixts already, that will load jemalloc and outputs the version or it fails to load jemalloc and outputs that "jemalloc library not found, memory usage may be reduced by installing it".

Let me know if this makes sense...

Cheers.

**Comment by amsmota** [comment]
@fab-10 , @usmansaleem et all,  I finally finsihed my changes, the failing test is now running and the other small changes done too. I tested again on my WSL and it all seems right.

Can youse please review the changes and hopefully test and give me some feedback? Please note that the expected scenarios that I tested are:

1. NO Jemalloc Installed / NO BESU_USING_JEMALLOC
2. NO Jemalloc Installed / BESU_USING_JEMALLOC = false
3. NO Jemalloc Installed / BESU_USING_JEMALLOC = true
4. WITH Jemalloc Installed / NO BESU_USING_JEMALLOC
5. WITH Jemalloc Installed / BESU_USING_JEMALLOC = false
6. WITH Jemalloc Installed / BESU_USING_JEMALLOC = true

and the results in the logs shloud be

1. jemalloc library not found, memory usage may be reduced by installing it
2. besu_using_jemalloc is present but is not set to true, jemalloc library not loaded
3. besu_using_jemalloc is present but we failed to load jemalloc library to get the version
4. jemalloc: 5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756
5. besu_using_jemalloc is present but is not set to true, jemalloc library not loaded
6. jemalloc: 5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa75

(jemalloc version can be diferent, of course)

Guys, I can't thank you enough for your help, If you ever come to Dublin (In Ireland, not Ohio, believe me or not Facebook already tagged me in Ohio! a few times!!!) I will gladly offer youse a couple of Guinnessess (it's a non written rule here that youse shall not drink only one).... 

Cheers.

File: CHANGELOG.md
Comment (amsmota) [reply CHANGELOG.md:105]: N/A

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java
Comment (fab-10) [inline besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1]: Please revert the changes not related to your work in this file

Comment (amsmota) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1]: Hmmm, what are those changes? I do see the Mockito there, is there any more?

I guess that some Optimize imports caused that... Anyway, I'll remove the Mockito's and compare...

Thanks for noticing....

Comment (fab-10) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1]: there are probably the result of a misresolved conflict during a merge from main, check also that the method `public void metricsUnknownCategoryRaiseError() {` has been removed on your PR but it should not

Comment (amsmota) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1]: I merged from the current main, the metricsUnknownCategoryRaiseError came from that merge, and all the Mockitos are gone.

Can your review again please?

Cheers.

**Review by fab-10 — APPROVED** [review]

**Comment by amsmota** [comment]
Hi @fab-10 , I see you aproved my last changes, can you tell me what are the next steps for me to be able to close this PR?

@usmansaleem I see one requested change made by you, can you approve/dismiss it?

Thank you guiys.

File: besu/src/main/scripts/unixStartScript.txt
Comment (usmansaleem) [inline besu/src/main/scripts/unixStartScript.txt:186]: @fab-10 is the intent correct? i.e. we only want to attempt to test and update LD_PRELOAD `libjemalloc` from startup script if and only if `BESU_USING_JEMALLOC` is not set to FALSE? 

My understanding was that we **always** attempt to load libjemalloc if we are able to test/load it via startup script and then set the variable BESU_USING_JEMALLOC so that we can log the JEMALLOC library version in our code at start up. (Because we never explicitly load jemalloc, it's just there available for JVM to use to allocate memory etc.)

I thought that `BESU_USING_JEMALLOC` is a misleading variable, it should be renamed to something like: `LOG_JEMALLOC_VERSION_AT_STARTUP` to clear the intent.

Anyways, I have removed my change request from the the PR, it is ready to merge. I'll leave it for you to proceed with the merge please.

Comment (fab-10) [reply besu/src/main/scripts/unixStartScript.txt:186]: yes this is now the intent, disable `jemalloc` if `BESU_USING_JEMALLOC=false`, so we keep the original auto-detect mechanism if the env var is not set.

@amsmota I will set auto merge on, please merge again with main

Comment (amsmota) [reply besu/src/main/scripts/unixStartScript.txt:186]: Hi guys, thanks for that. 

@usmansaleem, the objective of this PR was based on an issue we had when deploying in Prod some Besu nodes, deployment that succeed in the lower environments but not in Prod, so we had to remove Jemalloc and redeploy. I thougth then that if we had a env var to control the load/no-load it we could simply have set that env var to false and the deploy could go ahead without delays. And it ocurred to me that this could be used also for testing the performance gains of using Jemalloc, allowing for quickly testing the same app with our without it very quickly. Please note that if the env var does not exist at all Besu will behave exactly as it is now, it will output the current message if jemalloc does not exists ("jemalloc library not found, memory usage may be reduced by installing it") or goes ahead and prints the jemalloc version on the initial output of Besu.

@fab-10, so IIUC that means that I simply have to close the PR the merge will go ahead, so I'll do it now.

Again I would like to thank you guys for helping me understand the way jemalloc was loaded and reported on the logs, and changing my initial draft according to that.

Cheers guys.

**Comment by usmansaleem** [comment]
@amsmota You accidently closed it, I've reopened it and enabled automerge. What you need to do is merge the `main` branch back onto your branch. Once you have up-to-date changes from main branch, due to automerge enabled, it will merge once all tests pass.

**Comment by amsmota** [comment]
Hi @usmansaleem, thanks for your quick reply. I just did that merge so I'll wait for the tests to be run and close the PR.

Thank you again for you colaboration, it is greatly appreciated.

Cheers.

**Comment by usmansaleem** [comment]
@amsmota you don't need to "close" PR, once PR is merged to main, its status will automatically changed.

**Comment by amsmota** [comment]
> @amsmota you don't need to "close" PR, once PR is merged to main, its status will automatically changed.

So I can close the PR now?

**Comment by usmansaleem** [comment]
@amsmota As I said, if you close the PR it won't merge. As we have enabled "auto merge", it will merge on its own. Just sit back and relax :)

**Comment by amsmota** [comment]
> @amsmota As I said, if you close the PR it won't merge. As we have enabled "auto merge", it will merge on its own. Just sit back and relax :)

Got it 👍

**Comment by amsmota** [comment]
Guys, @usmansaleem and @fab-10 and everybody that helped with sugestions and comments my sincere gratitude, it was a task way more dificult than I initially supposed and I wouldn't be sucesffull in doing it without yours support. It seems that saying "standing in the shoulder of giants" it's actually true...

Well, "may the road rise to meet youse" as they say here in Ireland... 💯

**Pull request merged**
